### ISSUE TITLE
Improvement to keywords ?

### ISSUE THREAD

**Issue by eakteam** [CleverPumpkin/Proguard-Dictionaries-Generator#1]
Thanks for this great plugin, it really simplify things and save time.

I think that it would be better to add some like those `"reserved"` words to the dictionaries list, please take a look for better understand what i mean : 

https://github.com/facebookarchive/proguard/blob/master/examples/dictionaries/keywords.txt
https://github.com/facebookarchive/proguard/blob/master/examples/dictionaries/windows.txt

Example : 

If we will use this online tool http://www.javadecompilers.com/apk to try to decompile the universal APK obfuscated with dictionaries generated by this plugin it will `successfully decompile it`...

But if we use those above dictionaries it will `fail to decompile the APK`.

We have `tried on more online sites`, and seems to be the same, so it will be better to include those keywords for `better and more strong code protection`..

Thanks.

**Comment by creati8e** [comment]
Hi! This is a good suggestion. These keywords could be added to start of a dictionary file.

**Comment by eakteam** [comment]
Thanks for your reply.
Yes, this would be awesome. It will make the obfuscation stronger.

**Comment by creati8e** [comment]
It is interesting: i am trying to decompile apk obfuscated using reserved keywords and this site http://www.javadecompilers.com/apk replaces prohibited keywords
```
/* renamed from: volatile reason: invalid class name */
/* compiled from: ConstructorConstructor */
public class Cvolatile
```

**Comment by eakteam** [comment]
Yes, this is what i mean... If you use all reserved keywords like suggested above, it will not decompile it, also it shows errors in windows when trying to extract the archive, because it is not allowed in windows for some words :)

But normally if you use those above keywords as it should be that site will not continue to decompile it. It throws some errors

**Comment by creati8e** [comment]
It just replaces them

**Comment by eakteam** [comment]
It depends in the complexity of the code or how you implement the tests that you made to the test App. In my case the project is very large. About 8 On Demand Dynamic Features, Latest Play Core APIs and very much code. And in that case it doesn't success when trying to decompile. Not only that site but 4 other sites that i use for test and also offline version of JADX decompiler still doesn't works :D :D .... 

This is the app which i have tried with this plugin and with custom keywords. 
https://play.google.com/store/apps/details?id=com.eakteam.networkmanager.pro

With this tool every tool success decompiling it (maybe not as they should but they do it)
With custom keywords suggested in first comment, no tool is able to decompile it

**Comment by eakteam** [comment]
I use them in this order : 

```
#-obfuscationdictionary compact_keywords.txt
#-packageobfuscationdictionary keywords.txt
#-classobfuscationdictionary windows_keywords.txt
```

**Comment by creati8e** [comment]
I tried updated plugin on my production project with ~35 LoC. And it is decompiled succesfully, but Windows 10 is unable to open some files. Well, will continue experiments.
My approach for adding reserved keywords is following: i add them to the beginning of a dictionary file and them shuffle all lines in file.

**Comment by eakteam** [comment]
I understand. Please note that i use all three word lists in that repository and for 

`-obfuscationdictionary compact_keywords.txt`
`-packageobfuscationdictionary keywords.txt`
`-classobfuscationdictionary` using `windows_keywords.txt`

You said that 

> i add them to the beginning of a dictionary file and them shuffle all lines in file

So you mean that those keywords in word lists are being shuffled in the file ?
I don't know why you are able to decompile it , because i am just trying with my custom implementation and all those sites fail to decompile the app and throwing some weird errors : 

http://www.javadecompilers.com/apk
https://www.apkdecompilers.com/

This one fail too (but sometimes it success, but no matter that success it is unreadable)
http://apk-deguard.com/

**Comment by creati8e** [comment]
> Please note that i use all three word lists in that repository

Hmm, i used only -obfuscationdictionary and -classobfuscationdictionary. Maybe this is a cause.
>So you mean that those keywords in word lists are being shuffled in the file ?

Exactly.

**Comment by eakteam** [comment]
I don't know how the R8/Proguard parse exactly those words but i would suggest that no need to shuffle them, maybe use them more than once e.x after every 5K lines or 1K add same lines ???

Also try to use `-packageobfuscationdictionary with custom keywords.txt` to see if you will be unable to decompile. 

Also i would say that maybe no need to separate keywords for specific obfuscation. I mean join all the words from 3 files and use it to 3 types of obfuscation all of them ?

Sorry for my low English, hope to be understandable

In above repository they say : 

`KEYWORDS`

```
> #
> # This obfuscation dictionary contains reserved Java keywords. They can't
> # be used in Java source files, but they can be used in compiled class files.
> # Note that this hardly improves the obfuscation. Decent decompilers can
> # automatically replace reserved keywords, and the effect can fairly simply be
> # undone by obfuscating again with simpler names.
> # Usage:
> #     java -jar proguard.jar ..... -obfuscationdictionary keywords.txt
> #1 
```

`COMPAT`

```
> #
> # This obfuscation dictionary contains strings that are already present
> # in many class files. Since these strings can be shared, the resulting
> # obfuscated class files will generally be a little bit more compact.
> # Usage:
> #     java -jar proguard.jar ..... -obfuscationdictionary compact.txt
> #1 
```

`WINDOWS
`

```
> #
> # This obfuscation dictionary contains names that are not allowed as file names
> # in Windows, not even with extensions like .class or .java. They can however
> # be used without problems in jar archives, which just begs to apply them as
> # obfuscated class names. Trying to unpack the obfuscated archives in Windows
> # will probably generate some sparks.
> # Usage:
> #     java -jar proguard.jar ..... -classobfuscationdictionary windows.txt
> #                                  -packageobfuscationdictionary windows.txt
> #
```

**Comment by eakteam** [comment]
You should make about 13 tests to choose the right way how to implement they hahaha
I am sorry but this make the things better if you find the right way

**Comment by creati8e** [comment]
>join all the words from 3 files and use it

Yes, it do it in this way

**Comment by eakteam** [comment]
Hi @creati8e again. I know that maybe i have disturbed you, but yesterday you just solved the ugly issue with repository. But the latest version V. 1.0.7 it seems not to generate list of words from examples above in the generated dictionaries... Wish a great day to you and best regards.

Here i have attached the generated wordlist but no word from above examples inside. Hope to understand my English again.
[class-dictionary.txt](https://github.com/CleverPumpkin/Proguard-Dictionaries-Generator/files/3736636/class-dictionary.txt)
[obfuscation-dictionary.txt](https://github.com/CleverPumpkin/Proguard-Dictionaries-Generator/files/3736637/obfuscation-dictionary.txt)
[package-dictionary.txt](https://github.com/CleverPumpkin/Proguard-Dictionaries-Generator/files/3736638/package-dictionary.txt)

**Comment by creati8e** [comment]
Yes, you're right. When I've tested adding reserved words, it gave no effect to obfuscation hardness, so I decided to not make any changes related to this issue. But probably I should return to tests later.

**Comment by mrcas** [comment]
Is there anyway to use non characters in dictionary?
